### PR TITLE
typescript(vx-shape): Re-write package in TypeScript

### LINKED ISSUE #506: Re-write `@vx/shape` in TypeScript

### PR DESCRIPTION
**Pull request by williaster**
#### :rocket: Enhancements

This PR builds off #488 which introduces `Typescript` build config, and re-writes the `@vx/shape` package in `TypeScript`. Unfortunately this package is one of the biggest ones, so this PR is pretty sizable and larger than the other re-writes. 

A couple of notes:
- I added `@types/d3-path` and `@types/d3-shape` dependencies so that we can leverage/propagate existing types in our d3 usage.
- I removed the `@vx/point` dependency .
- I tried consolidating some of the duplicate code for the 12 link components, this was trickier than I thought so mostly share prop `Type`s.

#### :boom: Breaking Changes
- This PR introduces `React.Fragment`s, which requires bumping the `peerDep` for `react` to `^16.3.0-0`
- `@vx/shape` the `Arc` `centroid` prop was removed as it was not functional (it was called as if it was an `arc.centroid()` configuration parameter, but in reality the `.centroid` method is for returning the centroid of a datum.
-  `@vx/shape` the `Area` component is no longer wrapped in an empty `<g>` element
- `propTypes` across all components are likely stricter.
- `children(...)` render function overrides are wrapped in `React.Fragment`s  `<>{children(...)</>` to satisfy Types.
- `order` and `offset` props for `Stack`, `BarStack`, `BarStackHorizontal`, and `AreaStack` previously supported strings, arrays, or functions. Only the `string` prop was functional, and only the enumerated `string` presets are now allowed.
 
Closes #506

#### Testing
- [x] CI
  - [x] typescript
  - [x] eslint
  - [x] jest
- [x] Functional with `/gallery`
- [x] Generates `.d.ts` files

@hshoff @schillerk @milesj @techniq @kristw

### CODE REVIEW

File: packages/vx-shape/src/shapes/BarGroup.tsx [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+};
+
+/** One BarGroup is returned for each datum, which has multiple sub-bars (based on keys). */
+export interface BarGroup {
+  /** index of BarGroup (matches input Datum index). */
+  index: number;
+  /** x0 position of bar group */
+  x0: number;
+  /** bars within group, one for each key. */
+  bars: ({
```
Comment (Rudeg) [inline packages/vx-shape/src/shapes/BarGroup.tsx]: nit: maybe move `Bar` to the separate type, for the better readability?

File: packages/vx-shape/src/shapes/BarGroupHorizontal.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  /** @vx/scale or d3-scale that takes a group key and maps it to an y axis position (within a group). */
+  y1Scale: ScaleType;
+  /** Total width of the x-axis. */
+  width: number;
+  /** Override render function which is passed the computed Ba/rGroups. */
+  children?: (barGroups: BarGroup[]) => React.ReactNode;
+};
+
+/** One BarGroup is returned for each datum, which has multiple sub-bars (based on keys). */
+export interface BarGroup {
```
Comment (Rudeg) [inline packages/vx-shape/src/shapes/BarGroupHorizontal.tsx]: could you reuse `BarGroup` type from the `BarGroup.tsx` here?

Comment (williaster) [reply packages/vx-shape/src/shapes/BarGroupHorizontal.tsx]: they are slightly different, but I moved to a separate file and was able to consolidate `Bar` definitions (within a `BarGroup`) 👍

File: packages/vx-shape/src/shapes/link/curve/LinkHorizontalCurve.tsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  percent = 0.2,
+  x = (n: $TSFIXME) => n && n.y, // note this returns a y value
+  y = (n: $TSFIXME) => n && n.x, // note this returns an x value
+  source = (l: $TSFIXME) => l && l.source,
+  target = (l: $TSFIXME) => l && l.target,
+  ...restProps
+}: LinkHorizontalCurveProps<Link, Node> &
+  Omit<React.SVGProps<SVGPathElement>, keyof LinkHorizontalCurveProps<Link, Node>>) {
+  const pathGen = path || pathHorizontalCurve({ source, target, x, y, percent });
+  if (children) return <>{children({ path })}</>;
```
Comment (hshoff) [inline packages/vx-shape/src/shapes/link/curve/LinkHorizontalCurve.tsx]:
```jsx
if (children) return <>{children({ path: pathGen })}</>;
```

File: packages/vx-shape/src/shapes/link/curve/LinkRadialCurve.tsx [hunk 4]
```diff
@@ -36,34 +42,27 @@ export function pathRadialCurve({ source, target, x, y, percent }) {
...
+  y = (n: $TSFIXME) => n && n.y,
+  source = (l: $TSFIXME) => l && l.source,
+  target = (l: $TSFIXME) => l && l.target,
   ...restProps
-}) {
+}: LinkRadialCurveProps<Link, Node> &
+  Omit<React.SVGProps<SVGPathElement>, keyof LinkRadialCurveProps<Link, Node>>) {
   const pathGen = path || pathRadialCurve({ source, target, x, y, percent });
-  if (children) return children({ path });
+  if (children) return <>{children({ path })}</>;
```
Comment (hshoff) [inline packages/vx-shape/src/shapes/link/curve/LinkRadialCurve.tsx]:
```jsx
if (children) return <>{children({ path: pathGen })}</>;
```

File: packages/vx-shape/src/shapes/link/curve/LinkVerticalCurve.tsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  percent = 0.2,
+  x = (n: $TSFIXME) => n && n.x,
+  y = (n: $TSFIXME) => n && n.y,
+  source = (l: $TSFIXME) => l && l.source,
+  target = (l: $TSFIXME) => l && l.target,
+  ...restProps
+}: LinkVerticalCurveProps<Link, Node> &
+  Omit<React.SVGProps<SVGPathElement>, keyof LinkVerticalCurveProps<Link, Node>>) {
+  const pathGen = path || pathVerticalCurve({ source, target, x, y, percent });
+  if (children) return <>{children({ path })}</>;
```
Comment (hshoff) [inline packages/vx-shape/src/shapes/link/curve/LinkVerticalCurve.tsx]:
```jsx
if (children) return <>{children({ path: pathGen })}</>;
```

File: packages/vx-shape/src/shapes/link/diagonal/LinkRadial.tsx [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  path,
+  angle = (n: $TSFIXME) => n.x,
+  radius = (n: $TSFIXME) => n.y,
+  source = (n: $TSFIXME) => n.source,
+  target = (n: $TSFIXME) => n.target,
+  ...restProps
+}: LinkRadialDiagonalProps<Node, Link> &
+  Omit<React.SVGProps<SVGPathElement>, keyof LinkRadialDiagonalProps<Link, Node>>) {
+  const pathGen = path || pathRadialDiagonal({ source, target, angle, radius });
+  if (children) return <>{children({ path })}</>;
```
Comment (hshoff) [inline packages/vx-shape/src/shapes/link/diagonal/LinkRadial.tsx]:
```jsx
if (children) return <>{children({ path: pathGen })}</>;
```

File: packages/vx-shape/src/shapes/link/line/LinkHorizontalLine.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  path,
+  x = (d: $TSFIXME) => d.y, // note this returns a y value
+  y = (d: $TSFIXME) => d.x, // note this returns a y value
+  source = (d: $TSFIXME) => d.source,
+  target = (d: $TSFIXME) => d.target,
+  ...restProps
+}: LinkHorizontalLineProps<Link, Node> &
+  Omit<React.SVGProps<SVGPathElement>, keyof LinkHorizontalLineProps<Link, Node>>) {
+  const pathGen = path || pathHorizontalLine({ source, target, x, y });
+  if (children) return <>{children({ path })}</>;
```
Comment (hshoff) [inline packages/vx-shape/src/shapes/link/line/LinkHorizontalLine.tsx]:
```jsx
if (children) return <>{children({ path: pathGen })}</>;
```

File: packages/vx-shape/src/shapes/link/line/LinkRadialLine.tsx [hunk 8]
```diff
@@ -26,32 +31,23 @@ export function pathRadialLine({ source, target, x, y }) {
...
+  source = (d: $TSFIXME) => d.source,
+  target = (d: $TSFIXME) => d.target,
   children,
   ...restProps
-}) {
+}: LinkRadialLineProps<Link, Node> &
+  Omit<React.SVGProps<SVGPathElement>, keyof LinkRadialLineProps<Link, Node>>) {
   const pathGen = path || pathRadialLine({ source, target, x, y });
-  if (children) return children({ path });
+  if (children) return <>{children({ path })}</>;
```
Comment (hshoff) [inline packages/vx-shape/src/shapes/link/line/LinkRadialLine.tsx]:
```jsx
if (children) return <>{children({ path: pathGen })}</>;
```

File: packages/vx-shape/src/shapes/link/curve/LinkHorizontalCurve.tsx (hunk 3 above)
Comment (williaster) [reply packages/vx-shape/src/shapes/link/curve/LinkHorizontalCurve.tsx]: oof good catch

File: packages/vx-shape/src/util/setNumberOrNumberAccessor.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+/**
+ * This is a workaround for TypeScript not inferring the correct
+ * method overload/signature for some d3 shape methods.
+ */
+export default function setNumberOrNumberAccessor<NumAccessor>(
+  func: (d: number | NumAccessor) => void,
+  value: number | NumAccessor,
+) {
+  if (typeof value === 'number') func(value);
+  else func(value);
```
[the remark this answers is not in the document]
Comment (williaster) [reply packages/vx-shape/src/util/setNumberOrNumberAccessor.ts]: this one is so weird. asked mohsen about it and he wasn't sure 🤔

**Comment by williaster** [comment]
whew. lots of test errors that didn't always show up locally 🤔

**Pull request merged**